rbd: add check for imageFeatures parameter

Madhu-1 opened the pull request:
This commit adds checks for the missing `imageFeatures` parameter in createvolumerequest and nodestagerequest(only for static PVs). Missing `imageFeatures` parameter is ignored in the case of non-static PVs to ensure backward compatibility with older versions that did not have `imageFeatures` as a required parameter.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
rbd: add check for imageFeatures parameter

This commit adds checks for missing `imageFeatures` parameter
in createvolumerequest and nodestagerequest(only for static PVs).
Missing `imageFeatures` parameter is ignored in case of non-static
PVs to ensure backwards compatibility with older versions which
did not have `imageFeatures` as required parameter.

Signed-off-by: Rakshith R <rar@redhat.com>

Madhu-1 commented:
@Rakshith-R Thanks for the suggestion, Here is your PR to fix the problem.

mnaser commented:
I'm pretty sure this would still fail the test added here:

https://github.com/ceph/ceph-csi/pull/2135/commits/2cc7583a6ce2d86def26e6216d9769e8ee76b323

Madhu-1 commented:
@mnaser This will be merged once https://github.com/ceph/ceph-csi/pull/2135 gets merged. We need both PR's to fix the issue and provide a proper error message.

Rakshith-R commented:
> I'm pretty sure this would still fail the test added here:
> 
> [2cc7583](https://github.com/ceph/ceph-csi/commit/2cc7583a6ce2d86def26e6216d9769e8ee76b323)

@mnaser we still need the fix in your pr too, this is to have better error message for the users.

nixpanic commented:
I think there should be an e2e test that uses a static provisioned volume without the `imageFeatures` parameter. Can that be added?

Madhu-1 commented:
> I think there should be an e2e test that uses a static provisioned volume without the `imageFeatures` parameter. Can that be added?

@Rakshith-R incase if you missed ^^

Rakshith-R committed:
rbd: removing TrimSpace from validateImageFeatures func

`imageFeatures` string containing just whitespace should also
be treated as a invalid feature.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add e2e for static PVC without imageFeature parameter

This commit adds e2e to make sure static PVC without imageFeatures
parameter fail with proper error event.

Signed-off-by: Rakshith R <rar@redhat.com>

nixpanic commented on e2e/resize.go:
```diff
@@ -113,7 +113,7 @@ func resizePVCAndValidateSize(pvcPath, appPath string, f *framework.Framework) e
 	}
 	// wait for application pod to come up after resize
-	err = waitForPodInRunningState(app.Name, app.Namespace, f.ClientSet, deployTimeout)
+	err = waitForPodInRunningState(app.Name, app.Namespace, f.ClientSet, deployTimeout, "")
```
Instead of passing `""` as `expectedError`, define a const `noError = ""` and pass that. It will be much easier to understand when using the function in additional tests

Rakshith-R committed:
cleanup: added helper func isNotMountPoint

Added helper func isNotMountPoint to check mountPoint,
validate error and reduce complexity of NodeStageVolume.

Signed-off-by: Rakshith R <rar@redhat.com>

nixpanic commented on e2e/pod.go:
```diff
@@ -276,6 +284,19 @@ func waitForPodInRunningState(name, ns string, c kubernetes.Interface, t int) er
+		case v1.PodPending:
+			if expectedError != "" {
+				events, err := c.CoreV1().Events(ns).List(context.TODO(), metav1.ListOptions{
+					FieldSelector:fmt.Sprintf("involvedObject.name=%s",name)
```
I think this hasn't passed `go fmt`?
The line also needs to end in a `,`:
```
level=error msg="Running error: buildir: analysis skipped: errors in package: [/go/src/github.com/ceph/ceph-csi/e2e/pod.go:290:62: missing ',' before newline in composite literal]"
```

Rakshith-R replied:
fixed, thanks !

nixpanic commented on e2e/staticpvc.go:
```diff
@@ -77,7 +77,7 @@ func getStaticPVC(name, pvName, size, ns, sc string, blockPVC bool) *v1.Persiste
 }
 
-func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock bool) error {
+func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock, checkImgFeat bool) error {
```
This is getting ugly... more booleans as argument causes confusing what boolean is 1st/2nd.

Maybe we should start to introduce something like `checkFlags`:

```golang
const (
    defaultChecks = 0x0
    isBlock       = 0x1
    checkImgFeat  = 0x2
)

validateRBDStaticPV(f, rawAppPath, defaultChecks)
validateRBDStaticPV(f, rawAppPath, isBlock|checkImgFeat)
```

This probably makes the code easier to understand? Not needed to change now, but we have to start thinking about cleaning this up.

Rakshith-R replied:
true, i'll send a follow up pr to refractor/clean it up.

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented on e2e/staticpvc.go:
```diff
@@ -140,7 +142,11 @@ func validateRBDStaticPV(f *framework.Framework, appPath string, isBlock bool) e
 	app.Spec.Volumes[0].PersistentVolumeClaim.ClaimName = pvcName
-	err = createApp(f.ClientSet, app, deployTimeout)
+	if checkImgFeat {
+		err = createAppErr(f.ClientSet, app, deployTimeout, "missing required parameter imageFeatures")
```
keep the timeout less. so that CI can pass early? and make the error message we are checking as an argument.

Rakshith-R replied:
>keep the timeout less. so that CI can pass early? 

The function will exit as soon as the error msg is seen in the event i.e,. when nodeplugin responds for mount req, so it should'nt cause any delay for *passing* ci. 

>and make the error message we are checking as an argument.

I will send a follow up pr to cleanup and refactor functions better to address this and @nixpanic suggestion [here](https://github.com/ceph/ceph-csi/pull/2139#discussion_r655382522)

Rakshith-R commented:
rebased on top of devel PTAL @Madhu-1 @humblec @nixpanic @Yuggupta27

Madhu-1 commented:
LGTM

Madhu-1 commented:
As i cannot approve the PR am adding ready-to-merge label.

Madhu-1 commented:
@mergifyio rebase

Madhu-1 commented:
/test ci/centos/upgrade-tests-cephfs

Rakshith-R commented:
@humblec ,PTAL this is ready to merge/review .

humblec commented:
@Rakshith-R not in favor of  the `e2e` changes , however considering, i am completely refactoring the areas one by one, it will be picked  up soon refacoring, but we dont want to wait for that. Lets get this in :+1:

humblec commented:
@Mergifyio rebase

Rakshith-R commented:
rebased on top of devel, PTAL @humblec @nixpanic

Rakshith-R commented:
/retest ci/centos/upgrade-tests-rbd 
```

! StartHost failed, but will try again: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

* Creating kvm2 VM (CPUs=8, Memory=14336MB, Disk=20000MB) ...

* Failed to start kvm2 VM. Running "minikube delete" may fix it: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute



X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

```

Rakshith-R commented:
/retest ci/centos/upgrade-tests-rbd

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.19
```
+ ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@n24.pufty.ci.centos.org 'kubectl create namespace '\''cephcsi-e2e-6256571c3e2d'\'''

Warning: Permanently added 'n24.pufty.ci.centos.org,172.19.3.88' (ECDSA) to the list of known hosts.

Error from server: etcdserver: request timed out

script returned exit code 1
```

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.19

Failed due to #1969 ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.19/detail/mini-e2e-helm_k8s-1.19/1681/pipeline))

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.19

Rakshith-R commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.19

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.19/detail/mini-e2e-helm_k8s-1.19/1707/pipeline/107/
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.19/runs/1707/nodes/98/log/?start=0
Failed at
```
[1mSTEP[0m: create a Block mode PVC-PVC clone and bind it to an app
...
Jun 27 13:41:32.531: INFO: failed to delete PVC and application (rbd-6941): get on 
deleted PVC rbd-6941 failed with error other than "not found": Unauthorized
Jun 27 13:41:32.531: INFO: failed to delete PVC and application (rbd-6942): delete 
PV pvc-27cb1a73-99e5-4985-8df3-528d198b8539 failed with error other than "not found": Unauthorized
Jun 27 13:41:32.531: INFO: failed to delete PVC and application (rbd-6946): delete 
PV pvc-5c5d252a-4445-45db-bd47-144c367100e2 failed with error other than "not found": Unauthorized
Jun 27 13:41:32.531: FAIL: deleting PVCs and applications failed, 3 errors were logged
```
intermittent k8s api failure ?
